addons/mail/models/mail_link_preview.py

```diff
@@ -49,42 +45,38 @@ def _create_from_message_and_notify(self, message, request_url=None):
         for url in urls:
             if ignore_pattern and ignore_pattern.match(url):
                 continue
+            existing_link_preview = link_previews.search([("source_url", "=", url)])
+            if existing_link_preview and not url in link_previews_by_url:
+                self.env["mail.link.preview.message"].create([{
+                    "message_id": message.id,
+                    "link_preview_id": link_preview.id
+                } for link_preview in existing_link_preview])
+                link_previews += existing_link_preview
+                continue
             if url in link_previews_by_url:
                 preview = link_previews_by_url.pop(url)
-                if not preview.is_hidden:
-                    link_previews += preview
+                link_previews += preview
                 continue
             if preview := get_link_preview_from_url(url, requests_session):
-                preview['message_id'] = message.id
                 link_preview_values.append(preview)
             if len(link_preview_values) + len(link_previews) > 5:
                 break
-        for unused_preview in link_previews_by_url.values():
-            unused_preview._unlink_and_notify()
+        self.env["mail.link.preview.message"].search([
+            ("link_preview_id", "in", [unused_preview.id for unused_preview in link_previews_by_url.values()]),
```

Comment:
:pick: 

I feel like it's more verbose than it should be, if you replace this:

```diff
        link_previews_by_url = {
-           preview.source_url: preview for preview in message.sudo().link_preview_ids
+           preview.source_url: message.sudo().link_preview_ids
        }
```
Then you should be able to do:

```py
        self.env["mail.link.preview.message"].search([
            ("link_preview_id", "in", link_previews_by_url.values().ids),
            ...
```

Reply:
The loop is necessary to get `preview.source_url`.
Dict also doesn't seams to support `.ids` black magic: 
![image](https://github.com/user-attachments/assets/504e0c9c-14a1-42f4-9d02-ccf975263f42)